### PR TITLE
[nnc] Added a new API to distribute loop and all its parents

### PR DESCRIPTION
**Pull request by navahgar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61293**
* #61087
* #61085

Differential Revision: [D29560008](https://our.internmc.facebook.com/intern/diff/D29560008)

### CODE REVIEW

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29560008).

File: test/cpp/tensorexpr/test_loopnest.cpp [hunk 1]
```diff
@@ -5260,6 +5260,118 @@ TEST(LoopNest, DistributeLoopOverInnerLoops) {
...
+# CHECK-NEXT: for (int k
+# CHECK-NEXT: B[m, i] =
+# CHECK-NOT: for (
+        )IR";
+
+    auto newForI = dynamic_cast<For*>(Stmt::clone(forI));
+    auto forM = For::make(m, 0, 50, newForI);
+    auto par = Block::make({forM});
+    LoopNest nest(par, {a_buf.node(), b_buf.node()});
+    auto new_loops = LoopNest::distributeLoopAndParents(newForI);
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_loopnest.cpp:5332]: Nit: a combination of camelCase and snake_case :)

File: torch/csrc/jit/tensorexpr/loopnest.h [hunk 2]
```diff
@@ -238,6 +238,9 @@ class TORCH_API LoopNest {
   // S6:    for k
   // S7:      B[i] = B[i] +
   static std::vector<For*> distributeLoop(For* loop);
+  // Same as above, but also distribute parent loops.
+  // Returns the result of distributing the outermost loop.
+  static std::vector<For*> distributeLoopAndParents(For* loop);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/loopnest.h:243]: Could you please add an example of what it does? I have hard times understanding what exactly "also distribute parent loops" means.

**Review by ZolotukhinM — APPROVED** [review]
Overall looks good, I have only minor nits.

File: test/cpp/tensorexpr/test_loopnest.cpp (hunk 1 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_loopnest.cpp]: Fixed.

File: torch/csrc/jit/tensorexpr/loopnest.h (hunk 2 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/loopnest.h]: Added an example for these 2 APIs as well.

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29560008).

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29560008).

**Pull request closed without merging**